plugins/builds/index.js

```diff
@@ -785,18 +820,86 @@ const buildsPlugin = {
+            });
+
+            const flattenedWorkflowGraph = workflowParser.getFlattenedWorkflow(
+                event.workflowGraph,
+                stageWorkflowGraphs
+            );
             const current = {
                 pipeline,
                 job,
                 build,
-                event
+                event,
+                stageBuilds,
+                flattenedWorkflowGraph
             };
-
-            const nextJobsTrigger = workflowParser.getNextJobs(current.event.workflowGraph, {
+            let nextJobsTrigger = workflowParser.getNextJobs(current.flattenedWorkflowGraph, {
                 trigger: current.job.name,
                 chainPR: pipeline.chainPR
             });
 
+            const currentNode = current.flattenedWorkflowGraph.nodes.find(n => n.name === current.job.name);
+
+            // Figure out if need to run stage teardown
+            // If build is in stage
+            if (currentNode && currentNode.stageName) {
+                const currentStageBuild = stageBuilds.filter(sb => sb.stageName === currentNode.stageName)[0];
+
+                // Set stageBuild status to running if setup
+                if (currentNode.name === `stage@${currentNode.stageName}:setup`) {
+                    currentStageBuild.status = 'RUNNING';
+                    await currentStageBuild.update();
+                    // Set stageBuild status if completed
+                } else {
+                    //  else get all stage builds in workflow, compare number of stageBuilds to stage builds with status
+                    // Get job names from stage workflow
+                    const workflowStageBuilds = [];
+
+                    current.flattenedWorkflowGraph.nodes.forEach(n => {
+                        if (n.stageName === currentNode.stageName) {
+                            workflowStageBuilds.push(n.name);
+                        }
+                    });
+
+                    // get all builds in current stage
+                    const buildParentBuilds = current.build.parentBuilds;
+                    const stageDict = {};
+                    const eventBuilds = await current.event.getBuilds();
+
+                    if (buildParentBuilds && buildParentBuilds[pipeline.id]) {
+                        workflowStageBuilds.forEach(b => {
+                            if (b !== `stage@${currentNode.stageName}:teardown`) {
+                                const buildId = buildParentBuilds[pipeline.id].jobs[b];
+                                const buildForStage = eventBuilds.filter(eb => eb.id === buildId);
+
+                                stageDict[b] = buildForStage.length > 0 ? buildForStage[0] : undefined;
+                            }
+                        });
+                    }
+
+                    // Make sure number of builds is same as expected number of stage jobs
+                    if (Object.values(stageDict).every(b => b !== undefined)) {
```

Comment:
`teardown` may not get triggered in some scenarios.
For Ex: Below stage has 3 jobs
![image](https://github.com/screwdriver-cd/screwdriver/assets/2124590/922fae5a-ff76-4a00-9240-565072720a69)
When `prod-blue-test` fails, `teardown` should get triggered. But this logic expect build for all the jobs within a stage. Since `prod-blue-certify` does not have a build, `teardown` will not be triggered.